src/Interpreters/Cache/IFileCachePriority.h

```diff
@@ -6,49 +6,28 @@
 #include <Common/Exception.h>
 #include <Interpreters/Cache/FileCacheKey.h>
 #include <Interpreters/Cache/Guards.h>
+#include <Interpreters/Cache/FileCache_fwd_internal.h>
 
 namespace DB
 {
 
-class IFileCachePriority;
-using FileCachePriorityPtr = std::unique_ptr<IFileCachePriority>;
-struct KeyMetadata;
-using KeyMetadataPtr = std::shared_ptr<KeyMetadata>;
-
-namespace ErrorCodes
-{
-    extern const int LOGICAL_ERROR;
-}
-
 /// IFileCachePriority is used to maintain the priority of cached data.
-class IFileCachePriority
+class IFileCachePriority : private boost::noncopyable
 {
-friend class LockedCachePriority;
 public:
     using Key = FileCacheKey;
     using KeyAndOffset = FileCacheKeyAndOffset;
 
     struct Entry
     {
-        Entry(const Key & key_, size_t offset_, size_t size_, std::weak_ptr<KeyMetadata> key_metadata_)
-            : key(key_) , offset(offset_) , size(size_) , key_metadata(key_metadata_) {}
-
-        KeyMetadataPtr getKeyMetadata() const
-        {
-            auto result = key_metadata.lock();
-            if (!result)
-                throw Exception(ErrorCodes::LOGICAL_ERROR, "Metadata expired");
-            return result;
-        }
-
-        Key key;
-        size_t offset;
+        Entry(const Key & key_, size_t offset_, size_t size_, KeyMetadataPtr key_metadata_)
```

Comment:
`key_` param is redundant here, but in general, I wonder why we don't store `FileSegmentPtr` in entry. It'd be more natural.